### DISCUSSION TITLE
CoreWCF 0.4.0

### DISCUSSION THREAD

**Discussion by mconnew** [CoreWCF/CoreWCF#520] [Releases]
# Release Notes

## Blog Post
Blog post is available here: https://corewcf.github.io/blog/2021/12/15/corewcf-0_4_0_released

## Package Info

[CoreWCF.Primitives 0.4.0](https://www.nuget.org/packages/CoreWCF.Primitives/0.4.0)  
[CoreWCF.Http 0.4.0](https://www.nuget.org/packages/CoreWCF.Http/0.4.0)  
[CoreWCF.NetTcp 0.4.0](https://www.nuget.org/packages/CoreWCF.NetTcp/0.4.0)  
[CoreWCF.ConfigurationManager 0.4.0](https://www.nuget.org/packages/CoreWCF.ConfigurationManager/0.4.0)

## .NET Compatibility
Starting with 0.5.0, we have enabled tests to run for .NET 6. This means CoreWCF now supports .NET Core 3.1, .NET 5.0, .NET 6.0 and .NET Framework. We have dropped testing for .NET Core 2.1 as it is out of support. We still run tests against ASP.NET Core 2.1 as that is what runs on .NET Framework.

## New Features
* Added asynchronous support to SecurityTokenAuthenticator. - @g7ed6e
* Ported MTOM implementation from WCF Core client to provide MTOM message encoding support. - @g7ed6e
* Added support for scoped IServiceProvider available from OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>() - @g7ed6e
* Enabled ServiceSecurityContext.GetIdentities api including providing tests. - @JeffLeBert 
* Added MessageVersion values which were missing from .NET Framework. - @mconnew
* Added support to enable allowing DI to provide concrete type by only specifying an interface when configuring CoreWCF. - @Danielku15
* New source generator to enable easily adding DI injected parameters to service operations. - @g7ed6e

## Bug Fixes
* Fixed HttpRequest.PathBase being broken when middleware runs after the UseServiceModel middleware. - @mconnew
* Fixed unhandled exception in RequestReplyCorrelator during a race condition - @mconnew
* Fixed a Task which wasn't awaited in service dispatcher which could cause a service channel to be used before it was opened. Ported AsyncLocal for WCF Core client to fix broken implementation in CoreWCF. - @Danielku15
* Fixed FaultException<T> namespace not being correctly handled when using System.ServiceModel.FaultContext - @g7ed6e

<hr /><em>This discussion was created from the release <a href='https://github.com/CoreWCF/CoreWCF/releases/tag/v0.4.0'>CoreWCF 0.4.0</a>.</em>

**Comment by MeikTranel** [comment]
Nice, MTOM - 1 of 2 Blockers gone for me now. Still missing named pipes but this already unblocks some smaller migrations for me. Cheers to all contributors 🍻

**Reply by mconnew** [reply]
Named pipes will be coming next year.
